{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Day 2: SQANTI3 Practice Session\n",
    "see https://github.com/ConesaLab/SQANTI3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import os\n",
    "import subprocess\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize integrated IGV viewer in the notebook\n",
    "import igv_notebook\n",
    "igv_notebook.init()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installing SQANTI3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "sqanti_path = os.path.abspath(os.path.expanduser(\"~/tools/SQANTI3-5.4\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping SQANTI3 installation\n"
     ]
    }
   ],
   "source": [
    "%%script echo Skipping SQANTI3 installation\n",
    "# How to install SQANTI3?\n",
    "# see also: https://github.com/ConesaLab/SQANTI3/wiki/Dependencies-and-installation\n",
    "sqanti_zip = \"~/downloads/SQANTI3_v5.4.zip\"\n",
    "sqanti_env = \"sus_sqanti\"\n",
    "!wget https://github.com/ConesaLab/SQANTI3/releases/download/v5.4/SQANTI3_v5.4.zip -O $sqanti_zip\n",
    "!rm -rf $sqanti_path\n",
    "!mkdir -p $sqanti_path\n",
    "!unzip $sqanti_zip -d $sqanti_path\n",
    "!mv $sqanti_path/release_sqanti3/* $sqanti_path/\n",
    "!rm -r $sqanti_path/release_sqanti3\n",
    "# run in parent shell (outside of this notebook)\n",
    "# !micromamba env create -f $sqanti_path/SQANTI3.conda_env.yml -n $sqanti_env\n",
    "# !micromamba activate $sqanti_env\n",
    "# additional requirements for this course\n",
    "# !micromamba install -y ipykernel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparing paths to data ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_cores = 8\n",
    "data_dir = \"/mnt/c/data/lrgasp/homo_sapiens\"\n",
    "base_sus_path = os.path.abspath(os.path.expanduser(\"../../data\"))\n",
    "sample_name = \"h1_endo_chr8_isotools\"\n",
    "\n",
    "# reference\n",
    "ref_genome = f\"{data_dir}/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa\"\n",
    "ref_gtf = f\"{data_dir}/reference/chr8/gencode.v45.annotation_chr8.gtf\"\n",
    "ref_qc_prefix = \"gencode.v45.annotation_chr8_qc\"\n",
    "ref_qc_dir = f\"{base_sus_path}/sqanti_qc/gencode.v45.annotation_chr8\"\n",
    "\n",
    "# main data\n",
    "input_gtf = f\"{data_dir}/isotools/h1_endo_chr8_all.gtf\"\n",
    "qc_prefix = f\"{sample_name}_qc\"\n",
    "qc_dir = f\"{base_sus_path}/sqanti_qc/{sample_name}\"\n",
    "filter_prefix = f\"{sample_name}_filter\"\n",
    "filter_dir = f\"{base_sus_path}/sqanti_filter/{sample_name}\"\n",
    "rescue_prefix = f\"{sample_name}_rescue\"\n",
    "rescue_dir = f\"{base_sus_path}/sqanti_rescue/{sample_name}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# orthogonal data\n",
    "polyA_motifs = f\"{sqanti_path}/data/polyA_motifs/mouse_and_human.polyA_motif.txt\"\n",
    "cage = f\"{sqanti_path}/data/ref_TSS_annotation/human.refTSS_v3.1.hg38.bed\"\n",
    "counts = f\"{data_dir}/isotools/h1_endo_chr8_all_count.txt\"\n",
    "polyA_peaks = f\"{data_dir}/reference/orthogonal/chr8/atlas.clusters.2.0.GRCh38.96_chr8.bed\"\n",
    "splice_junctions = f\"{data_dir}/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPSJ.out_chr8.tab\"\n",
    "sr_bam = f\"{data_dir}/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPAligned.sortedByCoord.out_chr8.bam\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SQANTI3 Quality Control"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO: Motivate SQANTI3 QC\n",
    "Copy some basic information, graphs, etc. about SQANTI from the docs and tutorials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    =====================\n",
      "      ░██████╗░░█████╗░\n",
      "      ██╔═══██╗██╔══██╗\n",
      "      ██║██╗██║██║░░╚═╝\n",
      "      ╚██████╔╝██║░░██╗\n",
      "      ░╚═██╔═╝░╚█████╔╝\n",
      "      ░░░╚═╝░░░░╚════╝░\n",
      "    =====================\n",
      "    \n",
      "usage: sqanti3_qc.py [-h] --isoforms ISOFORMS --refGTF REFGTF --refFasta\n",
      "                     REFFASTA [--min_ref_len MIN_REF_LEN] [--force_id_ignore]\n",
      "                     [--fasta] [--genename]\n",
      "                     [--novel_gene_prefix NOVEL_GENE_PREFIX] [-s SITES]\n",
      "                     [-w WINDOW]\n",
      "                     [--aligner_choice {minimap2,deSALT,gmap,uLTRA}]\n",
      "                     [-x GMAP_INDEX] [--skipORF] [--orf_input ORF_INPUT]\n",
      "                     [--short_reads SHORT_READS] [--SR_bam SR_BAM]\n",
      "                     [--CAGE_peak CAGE_PEAK]\n",
      "                     [--polyA_motif_list POLYA_MOTIF_LIST]\n",
      "                     [--polyA_peak POLYA_PEAK] [--phyloP_bed PHYLOP_BED]\n",
      "                     [-e EXPRESSION] [-c COVERAGE] [-fl FL_COUNT]\n",
      "                     [--isoAnnotLite] [--gff3 GFF3] [-o OUTPUT] [-d DIR]\n",
      "                     [--saturation] [--report {html,pdf,both,skip}]\n",
      "                     [--isoform_hits]\n",
      "                     [--ratio_TSS_metric {max,mean,median,3quartile}]\n",
      "                     [-t CPUS] [-n CHUNKS] [-l {ERROR,WARNING,INFO,DEBUG}]\n",
      "                     [--is_fusion] [-v] [--bugsi {human,mouse}]\n",
      "\n",
      "Structural and Quality Annotation of Novel Transcript Isoforms\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Required arguments:\n",
      "  --isoforms ISOFORMS   Isoforms (FASTA/FASTQ) or GTF format. It is\n",
      "                        recommended to provide them in GTF format, but if it\n",
      "                        is needed to map the sequences to the genome use a\n",
      "                        FASTA/FASTQ file with the --fasta option.\n",
      "  --refGTF REFGTF       Reference annotation file (GTF format)\n",
      "  --refFasta REFFASTA   Reference genome (Fasta format)\n",
      "\n",
      "Customization and filtering:\n",
      "  --min_ref_len MIN_REF_LEN\n",
      "                        Minimum reference transcript length (default: 0 bp)\n",
      "  --force_id_ignore     Allow the usage of transcript IDs non related with\n",
      "                        PacBio's nomenclature (PB.X.Y)\n",
      "  --fasta               Use when running SQANTI by using as input a\n",
      "                        FASTA/FASTQ with the sequences of isoforms\n",
      "  --genename            Use gene_name tag from GTF to define genes. Default:\n",
      "                        gene_id used to define genes\n",
      "  --novel_gene_prefix NOVEL_GENE_PREFIX\n",
      "                        Prefix for novel isoforms (default: None)\n",
      "  -s SITES, --sites SITES\n",
      "                        Set of splice sites to be considered as canonical, in\n",
      "                        a comma separated list. (default: ATAC,GCAG,GTAG)\n",
      "  -w WINDOW, --window WINDOW\n",
      "                        Size of the window in the genomic DNA screened for\n",
      "                        Adenine content downstream of TTS (default: 20)\n",
      "\n",
      "Aligner and mapping options:\n",
      "  --aligner_choice {minimap2,deSALT,gmap,uLTRA}\n",
      "                        Select your aligner of choice: minimap2, deSALT, gmap,\n",
      "                        uLTRA (default: minimap2)\n",
      "  -x GMAP_INDEX, --gmap_index GMAP_INDEX\n",
      "                        Path and prefix of the reference index created by\n",
      "                        gmap_build. Mandatory if using GMAP .\n",
      "\n",
      "ORF prediction:\n",
      "  --skipORF             Skip ORF prediction (to save time)\n",
      "  --orf_input ORF_INPUT\n",
      "                        Input fasta to run ORF on. By default, ORF is run on\n",
      "                        genome-corrected fasta - this overrides it. If input\n",
      "                        is fusion (--is_fusion), this must be provided for ORF\n",
      "                        prediction.\n",
      "\n",
      "Orthogonal data inputs:\n",
      "  --short_reads SHORT_READS\n",
      "                        File Of File Names (fofn, space separated) with paths\n",
      "                        to FASTA or FASTQ from Short-Read RNA-Seq. If\n",
      "                        expression or coverage files are not provided,\n",
      "                        Kallisto (just for pair-end data) and STAR,\n",
      "                        respectively, will be run to calculate them.\n",
      "  --SR_bam SR_BAM       Directory or fofn file with the sorted bam files of\n",
      "                        Short Reads RNA-Seq mapped against the genome\n",
      "  --CAGE_peak CAGE_PEAK\n",
      "                        FANTOM5 Cage Peak (BED format, optional)\n",
      "  --polyA_motif_list POLYA_MOTIF_LIST\n",
      "                        Ranked list of polyA motifs (text, optional)\n",
      "  --polyA_peak POLYA_PEAK\n",
      "                        PolyA Peak (BED format, optional)\n",
      "  --phyloP_bed PHYLOP_BED\n",
      "                        PhyloP BED for conservation score (BED, optional)\n",
      "  -e EXPRESSION, --expression EXPRESSION\n",
      "                        Expression matrix (supported: Kallisto tsv)\n",
      "  -c COVERAGE, --coverage COVERAGE\n",
      "                        Junction coverage files (provide a single file, comma-\n",
      "                        delmited filenames, or a file pattern, ex:\n",
      "                        \"mydir/*.junctions\").\n",
      "  -fl FL_COUNT, --fl_count FL_COUNT\n",
      "                        Full-length PacBio abundance file\n",
      "\n",
      "Functional annotation:\n",
      "  --isoAnnotLite        Run isoAnnot Lite to output a tappAS-compatible gff3\n",
      "                        file\n",
      "  --gff3 GFF3           Precomputed tappAS species specific GFF3 file. It will\n",
      "                        serve as reference to transfer functional attributes\n",
      "\n",
      "Output options:\n",
      "  -o OUTPUT, --output OUTPUT\n",
      "                        Prefix for output files\n",
      "  -d DIR, --dir DIR     Directory for output files. (Default: Directory where\n",
      "                        the script was run.)\n",
      "  --saturation          Include saturation curves into report\n",
      "  --report {html,pdf,both,skip}\n",
      "                        Select report format: html, pdf, both, skip (default:\n",
      "                        html)\n",
      "  --isoform_hits        Report all FSM/ISM isoform hits in a separate file\n",
      "  --ratio_TSS_metric {max,mean,median,3quartile}\n",
      "                        Define which statistic metric should be reported in\n",
      "                        the ratio_TSS column (default: max)\n",
      "\n",
      "Performance options:\n",
      "  -t CPUS, --cpus CPUS  Number of threads used during alignment by aligners.\n",
      "                        (default: 10)\n",
      "  -n CHUNKS, --chunks CHUNKS\n",
      "                        Number of chunks to split SQANTI3 analysis in for\n",
      "                        speed up (default: 1).\n",
      "  -l {ERROR,WARNING,INFO,DEBUG}, --log_level {ERROR,WARNING,INFO,DEBUG}\n",
      "                        Set the logging level INFO\n",
      "\n",
      "Optional arguments:\n",
      "  --is_fusion           Input are fusion isoforms, must supply GTF as input\n",
      "  -v, --version         Display program version number.\n",
      "  --bugsi {human,mouse}\n",
      "                        Generate a BUGSI benchmarking report for the given\n",
      "                        species (human or mouse)\n"
     ]
    }
   ],
   "source": [
    "# How to run SQANTI3 Quality Control?\n",
    "# see also: https://github.com/ConesaLab/SQANTI3/wiki/Running-SQANTI3-Quality-Control#running\n",
    "!$sqanti_path/sqanti3_qc.py --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running SQ3 Quality Control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command:\n",
      "/usr/bin/time -v /home/fjetzinger/tools/SQANTI3-5.4/sqanti3_qc.py --isoforms /mnt/c/data/lrgasp/homo_sapiens/isotools/h1_endo_chr8_all.gtf --refGTF /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/gencode.v45.annotation_chr8.gtf --refFasta /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa --polyA_motif_list /home/fjetzinger/tools/SQANTI3-5.4/data/polyA_motifs/mouse_and_human.polyA_motif.txt --polyA_peak /mnt/c/data/lrgasp/homo_sapiens/reference/orthogonal/chr8/atlas.clusters.2.0.GRCh38.96_chr8.bed --coverage /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPSJ.out_chr8.tab --CAGE_peak /home/fjetzinger/tools/SQANTI3-5.4/data/ref_TSS_annotation/human.refTSS_v3.1.hg38.bed --SR_bam /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPAligned.sortedByCoord.out_chr8.bam --output h1_endo_chr8_isotools_qc --dir /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools --skipORF --cpus 8\n",
      "Output:\n",
      "\n",
      "Errors:\n",
      "\n",
      "    =====================\n",
      "      ░██████╗░░█████╗░\n",
      "      ██╔═══██╗██╔══██╗\n",
      "      ██║██╗██║██║░░╚═╝\n",
      "      ╚██████╔╝██║░░██╗\n",
      "      ░╚═██╔═╝░╚█████╔╝\n",
      "      ░░░╚═╝░░░░╚════╝░\n",
      "    =====================\n",
      "    \n",
      "[INFO:2025-05-14 15:13:58,755] Write arguments to /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc.qc_params.txt...\n",
      "[INFO:2025-05-14 15:13:58,769] Initialising QC pipeline.\n",
      "[INFO:2025-05-14 15:13:58,770] Parsing provided files\n",
      "[INFO:2025-05-14 15:13:58,771] Reading genome fasta /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa\n",
      "[INFO:2025-05-14 15:14:08,663] **** Correcting sequences\n",
      "[INFO:2025-05-14 15:14:08,665] Correcting fasta\n",
      "[INFO:2025-05-14 15:14:08,665] Skipping aligning of sequences because GTF file was provided.\n",
      "[INFO:2025-05-14 15:14:14,834] Indels will be not calculated since you ran SQANTI3 without alignment step (SQANTI3 with gtf format as transcriptome input).\n",
      "[INFO:2025-05-14 15:14:43,109] **** Predicting ORF sequences...\n",
      "[WARNING:2025-05-14 15:14:43,118] Skipping ORF prediction because user requested it. All isoforms will be non-coding!\n",
      "[WARNING:2025-05-14 15:14:43,119] All input isoforms were predicted as non-coding\n",
      "[INFO:2025-05-14 15:14:43,120] **** Parsing Reference Transcriptome....\n",
      "[INFO:2025-05-14 15:14:47,620] **** Parsing Isoforms.\n",
      "[INFO:2025-05-14 15:14:48,776] Input pattern: /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPSJ.out_chr8.tab.\n",
      "The following files found and to be read as junctions:\n",
      "/mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPSJ.out_chr8.tab\n",
      "[INFO:2025-05-14 15:14:48,840] 10577 junctions read. 14 junctions added to both strands because no strand information from STAR.\n",
      "[INFO:2025-05-14 15:14:48,841] Using provided BAM files for calculating TSS ratio\n",
      "[INFO:2025-05-14 15:14:54,650] BAM files identified: ['/mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPAligned.sortedByCoord.out_chr8.bam']\n",
      "[INFO:2025-05-14 15:15:38,756] **** Reading CAGE Peak data.\n",
      "[INFO:2025-05-14 15:15:39,568] **** Reading polyA Peak data.\n",
      "[INFO:2025-05-14 15:15:40,149] **** Reading PolyA motif list.\n",
      "[INFO:2025-05-14 15:15:40,152] **** Performing Classification of Isoforms\n",
      "[INFO:2025-05-14 15:18:36,921] Number of classified isoforms: 22417\n",
      "[INFO:2025-05-14 15:18:37,132] After classify fsm: 22417\n",
      "[INFO:2025-05-14 15:18:37,148] Full-length read abundance files not provided.\n",
      "[INFO:2025-05-14 15:18:37,154] RT-switching computation\n",
      "[INFO:2025-05-14 15:18:42,835] After RTS classificaion: 22417\n",
      "[INFO:2025-05-14 15:18:42,836] **** Adding TSS ratio data.\n",
      "[INFO:2025-05-14 15:18:42,871] Isoforms expression files not provided.\n",
      "[INFO:2025-05-14 15:18:46,611] Writing output files\n",
      "[INFO:2025-05-14 15:19:18,018] **** Generating SQANTI3 report.\n",
      "[INFO:2025-05-14 15:20:21,538] Removing temporary files.\n",
      "[INFO:2025-05-14 15:20:21,546] SQANTI3 complete in 382.77151822007727 sec.\n",
      "\tCommand being timed: \"/home/fjetzinger/tools/SQANTI3-5.4/sqanti3_qc.py --isoforms /mnt/c/data/lrgasp/homo_sapiens/isotools/h1_endo_chr8_all.gtf --refGTF /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/gencode.v45.annotation_chr8.gtf --refFasta /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa --polyA_motif_list /home/fjetzinger/tools/SQANTI3-5.4/data/polyA_motifs/mouse_and_human.polyA_motif.txt --polyA_peak /mnt/c/data/lrgasp/homo_sapiens/reference/orthogonal/chr8/atlas.clusters.2.0.GRCh38.96_chr8.bed --coverage /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPSJ.out_chr8.tab --CAGE_peak /home/fjetzinger/tools/SQANTI3-5.4/data/ref_TSS_annotation/human.refTSS_v3.1.hg38.bed --SR_bam /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPAligned.sortedByCoord.out_chr8.bam --output h1_endo_chr8_isotools_qc --dir /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools --skipORF --cpus 8\"\n",
      "\tUser time (seconds): 259.19\n",
      "\tSystem time (seconds): 32.78\n",
      "\tPercent of CPU this job got: 75%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 6:29.11\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 589140\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 1898\n",
      "\tMinor (reclaiming a frame) page faults: 455595\n",
      "\tVoluntary context switches: 197662\n",
      "\tInvoluntary context switches: 43696\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 404040\n",
      "\tFile system outputs: 19664\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Build the SQANTI3 QC command\n",
    "cmd = [\n",
    "    \"/usr/bin/time\", \"-v\",\n",
    "    f\"{sqanti_path}/sqanti3_qc.py\",         # SQANTI3 QC script\n",
    "    \"--isoforms\", input_gtf,                # GTF to Quality Control\n",
    "    \"--refGTF\", ref_gtf,                    # Reference GTF\n",
    "    \"--refFasta\", ref_genome,               # Reference Genome\n",
    "    \"--polyA_motif_list\", polyA_motifs,     # PolyA Motif List\n",
    "    \"--polyA_peak\", polyA_peaks,            # PolyA Peaks\n",
    "    # \"--fl_count\", counts,                   # Counts file\n",
    "    \"--coverage\", splice_junctions,         # splice junction short-read coverage file (from STAR)\n",
    "    \"--CAGE_peak\", cage,                    # CAGE Peak file\n",
    "    \"--SR_bam\", sr_bam,                     # Short-read BAM file\n",
    "    \"--output\", qc_prefix,                  # Output Prefix\n",
    "    \"--dir\", qc_dir,                        # Output Location\n",
    "    \"--skipORF\",                            # Skip ORF Prediction (takes longer)\n",
    "    \"--cpus\", str(n_cores)                  # Number of Threads\n",
    "]\n",
    "\n",
    "# Print the command for reference\n",
    "print(\"Running command:\")\n",
    "print(\" \".join(cmd))\n",
    "\n",
    "# Run the command\n",
    "result = subprocess.run(cmd, capture_output=True, text=True)\n",
    "\n",
    "# Print output and errors\n",
    "print(\"Standard Output:\")\n",
    "print(result.stdout)\n",
    "print(\"Standard Error:\")\n",
    "print(result.stderr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Investigating SQANTI3 QC results\n",
    "\n",
    "see also: https://github.com/ConesaLab/SQANTI3/wiki/Understanding-the-output-of-SQANTI3-QC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Can you find answers to the following Questions in the SQANTI3 Quality Control report?\n",
    "\n",
    "SQANTI3 provides a detailed report with a great variety of information, statistics, and plots that detail the results of the Quality Control process. You can find it under `data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_SQANTI3_report.html`\n",
    "\n",
    "1. **Splice Junction Classification:**\n",
    "    \n",
    "    a. How many non-canonical splice junctions exist in the transcriptome? How many of them are known vs. novel?\n",
    "\n",
    "    b. Which structural categories do you think the majority of transcripts with novel non-canonical splice junctions belong to? \n",
    "\n",
    "2. **Exon Structure:**\n",
    "\n",
    "    a. How many transcripts are multi- vs. mono-exonic? How does this differ between reference and novel transcripts?\n",
    "\n",
    "    b. Can you think of some possible reasons for novel transcripts to contain more mono-exons than reference transcripts?\n",
    "\n",
    "3. **Features of Good Quality:**\n",
    "\n",
    "    a. Examine how the support levels of annotation, canonical splice junctions, splice junction coverage, and CAGE coverage differ across the FSM, ISM, NIC, and NNC structural categories. Then, try to think of possible explanations for what you are observing.\n",
    "\n",
    "    b. Can you explain the following observations?\n",
    "\n",
    "    1. ISM transcripts show noticably lower levels of CAGE support.\n",
    "\n",
    "    2. NNC transcripts show noticably lower levels of Canonical splice junctions as well as splice junction coverage.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Investigate in Genome Viewer..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div id=\"jb_1630860_buttons\"></div><div id=\"jb_1630860_igvcontainer\"></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "window.igv.MessageHandler.on({\"id\": \"jb_1630860\", \"command\": \"createBrowser\", \"data\": {\"reference\": {\"id\": \"hg38\", \"name\": \"Human (GRCH38/hg38)\", \"fastaPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa\", \"indexPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa.fai\"}, \"locus\": \"chr8:75,391,802-75,577,425\", \"tracks\": [{\"name\": \"IsoTools\", \"path\": \"../../data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_corrected.gtf\", \"format\": \"gtf\", \"type\": \"annotation\", \"displayMode\": \"SQUISHED\"}, {\"name\": \"Reference\", \"path\": \"../../data/sqanti_qc/gencode.v45.annotation_chr8/gencode.v45.annotation_chr8_qc_corrected.gtf\", \"format\": \"gtf\", \"type\": \"annotation\", \"displayMode\": \"SQUISHED\"}], \"id\": \"jb_1630860\"}})",
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "igv_browser_qc= igv_notebook.Browser(\n",
    "    {\n",
    "        \"reference\": {\n",
    "            \"id\": \"hg38\",\n",
    "            \"name\": \"Human (GRCH38/hg38)\",\n",
    "            \"fastaPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa\",\n",
    "            \"indexPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa.fai\"\n",
    "        },\n",
    "        \"locus\": \"chr8:75,391,802-75,577,425\",\n",
    "        \"tracks\": [\n",
    "            {\n",
    "                \"name\": \"IsoTools\",\n",
    "                \"path\": \"../../data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_corrected.gtf\",\n",
    "                \"format\": \"gtf\",\n",
    "                \"type\": \"annotation\",\n",
    "                \"displayMode\": \"SQUISHED\"\n",
    "            },\n",
    "            {\n",
    "            \"name\": \"Reference\",\n",
    "            \"path\": \"../../data/sqanti_qc/gencode.v45.annotation_chr8/gencode.v45.annotation_chr8_qc_corrected.gtf\",\n",
    "            \"format\": \"gtf\",\n",
    "            \"type\": \"annotation\",\n",
    "            \"displayMode\": \"SQUISHED\"\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SQANTI3 Filter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO: Motivate SQANTI3 Filter\n",
    "Copy some basic information, graphs, etc. about SQANTI from the docs and tutorials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    ================================================     \n",
      "      ███████╗██╗██╗░░░░░████████╗███████╗██████╗░\n",
      "      ██╔════╝██║██║░░░░░╚══██╔══╝██╔════╝██╔══██╗\n",
      "      █████╗░░██║██║░░░░░░░░██║░░░█████╗░░██████╔╝\n",
      "      ██╔══╝░░██║██║░░░░░░░░██║░░░██╔══╝░░██╔══██╗\n",
      "      ██║░░░░░██║███████╗░░░██║░░░███████╗██║░░██║\n",
      "      ╚═╝░░░░░╚═╝╚══════╝░░░╚═╝░░░╚══════╝╚═╝░░╚═╝\n",
      "    ================================================\n",
      "    \n",
      "usage: sqanti3_filter.py ml [-h] --sqanti_class SQANTI_CLASS\n",
      "                            [--isoAnnotGFF3 ISOANNOTGFF3]\n",
      "                            [--filter_isoforms FILTER_ISOFORMS]\n",
      "                            [--filter_gtf FILTER_GTF]\n",
      "                            [--filter_sam FILTER_SAM]\n",
      "                            [--filter_faa FILTER_FAA] [-o OUTPUT] [-d DIR]\n",
      "                            [--skip_report] [-e] [-v] [-c CPUS]\n",
      "                            [-l {ERROR,WARNING,INFO,DEBUG}]\n",
      "                            [-t PERCENT_TRAINING] [-p TP] [-n TN]\n",
      "                            [-j THRESHOLD] [-f] [--intermediate_files]\n",
      "                            [-r REMOVE_COLUMNS] [-z MAX_CLASS_SIZE]\n",
      "                            [-i INTRAPRIMING]\n",
      "\n",
      "ML filter selected\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Required arguments:\n",
      "  --sqanti_class SQANTI_CLASS\n",
      "                        SQANTI3 QC classification file.\n",
      "\n",
      "Input options:\n",
      "  --isoAnnotGFF3 ISOANNOTGFF3\n",
      "                        isoAnnotLite GFF3 file to be filtered\n",
      "  --filter_isoforms FILTER_ISOFORMS\n",
      "                        fasta/fastq isoform file to be filtered\n",
      "  --filter_gtf FILTER_GTF\n",
      "                        GTF file to be filtered\n",
      "  --filter_sam FILTER_SAM\n",
      "                        SAM alignment of the input fasta/fastq\n",
      "  --filter_faa FILTER_FAA\n",
      "                        ORF prediction faa file to be filtered by SQANTI3\n",
      "\n",
      "Output options:\n",
      "  -o OUTPUT, --output OUTPUT\n",
      "                        Prefix for output files.\n",
      "  -d DIR, --dir DIR     Directory for output files. Default: ./sqanti3_results\n",
      "  --skip_report         Skip creation of a report about the filtering\n",
      "\n",
      "Filtering options:\n",
      "  -e, --filter_mono_exonic\n",
      "                        All mono-exonic transcripts are automatically filtered\n",
      "\n",
      "Extra options:\n",
      "  -v, --version         Display program version number.\n",
      "  -c CPUS, --cpus CPUS  Number of CPUs to use. Default: 4\n",
      "  -l {ERROR,WARNING,INFO,DEBUG}, --log_level {ERROR,WARNING,INFO,DEBUG}\n",
      "                        Set the logging level.\n",
      "                        Default: INFO\n",
      "\n",
      "Machine Learning specific options:\n",
      "  -t PERCENT_TRAINING, --percent_training PERCENT_TRAINING\n",
      "                        Proportion of the data that goes to training (parameter p of the function createDataPartition).                     \n",
      "                        Default: 0.8\n",
      "  -p TP, --TP TP        Path to file containing the list of the TP transcripts, one ID by line, no header (optional). If not supplied, it will be generated from input data.\n",
      "  -n TN, --TN TN        Path to file containing the list of the TN transcripts, one ID by line, no header (optional). If not supplied, it will be generated from input data.\n",
      "  -j THRESHOLD, --threshold THRESHOLD\n",
      "                        Machine Learning probability threshold to classify transcripts as positive isoforms.         \n",
      "                        Default: 0.7\n",
      "  -f, --force_fsm_in    Forces retaining FMS transcripts regardless of ML filter result (FSM are threfore automatically classified as isoforms).\n",
      "  --intermediate_files  Outputs ML filter intermediate files.\n",
      "  -r REMOVE_COLUMNS, --remove_columns REMOVE_COLUMNS\n",
      "                        Path to single-column file (no header) containing the names of the columns in SQ3's classification.txt file that are to be excluded during random forest training (optional).\n",
      "  -z MAX_CLASS_SIZE, --max_class_size MAX_CLASS_SIZE\n",
      "                        Maximum number of isoforms to include in True Positive and True Negative sets. TP and TN sets will be downsized to this value if they are larger.\n",
      "                        Default: 3000\n",
      "  -i INTRAPRIMING, --intrapriming INTRAPRIMING\n",
      "                        Adenine percentage at genomic 3' end to flag an isoform as intra-priming. Default: 60\n"
     ]
    }
   ],
   "source": [
    "# How to run SQ3 Filter?\n",
    "# see also: https://github.com/ConesaLab/SQANTI3/wiki/Running-SQANTI3-filter\n",
    "# specifically for the Machine Learning Filter: https://github.com/ConesaLab/SQANTI3/wiki/Running-SQANTI3-filter#ml\n",
    "!$sqanti_path/sqanti3_filter.py ml --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running SQ3 Filter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocessing: Choosing True Positive / True Negative Transcript Sets \n",
    "While this can be done automatically, we have found the following set of criteria to yield more stringent filtering results.\n",
    "The SQANTI3 Machine Learning Filter then trains a Random Forest model on the basis of these True Positive / True Negative transcripts, which is then applied to filter the entire transcriptome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved 1036 TP isoforms to /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/filter_sets/TP_list.txt\n",
      "Saved 3000 TN isoforms to /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/filter_sets/TN_list.txt\n",
      "Exclusion list saved to /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/filter_sets/exclusion_list.txt\n"
     ]
    }
   ],
   "source": [
    "classification_file = os.path.join(qc_dir, f\"{qc_prefix}_classification.txt\")\n",
    "filter_sets_dir = os.path.join(filter_dir, \"filter_sets\")\n",
    "max_set_size = 3000\n",
    "\n",
    "# Load classification file\n",
    "df = pd.read_csv(classification_file, sep='\\t', low_memory=False)\n",
    "\n",
    "# Ensure required columns are present\n",
    "required_columns = ['structural_category', 'all_canonical', 'within_CAGE_peak', 'within_polyA_site', 'isoform']\n",
    "for col in required_columns:\n",
    "    if col not in df.columns:\n",
    "        raise ValueError(f\"Required column '{col}' not found in the classification file.\")\n",
    "\n",
    "# Define True Positives (TP)\n",
    "tp_conditions = (\n",
    "    (df['structural_category'] == 'full-splice_match') &\n",
    "    (df['all_canonical'] == \"canonical\") &\n",
    "    (df['within_CAGE_peak'] == True) &\n",
    "    (df['within_polyA_site'] == True) &\n",
    "    (df['exons'] > 1) \n",
    ")\n",
    "tp_df = df[tp_conditions]\n",
    "\n",
    "    # Define True Negatives (TN)\n",
    "tn_conditions = (\n",
    "    (df['structural_category'] != 'full-splice_match') &\n",
    "    (\n",
    "        (df['all_canonical'] == \"non_canonical\") |\n",
    "        (df['within_CAGE_peak'] != True) |\n",
    "        (df['within_polyA_site'] != True)\n",
    "    ) &\n",
    "    (df['exons'] > 1)\n",
    ")\n",
    "tn_df = df[tn_conditions]\n",
    "\n",
    "# Sample up to max_size transcripts for each set\n",
    "tp_sample = tp_df.sample(n=min(len(tp_df), max_set_size), random_state=42)\n",
    "tn_sample = tn_df.sample(n=min(len(tn_df), max_set_size), random_state=42)\n",
    "\n",
    "if len(tp_sample) < 250:\n",
    "    raise ValueError(f\"Not enough TP transcripts found. Found {len(tp_sample)}, required 250.\")\n",
    "if len(tn_sample) < 250:\n",
    "    raise ValueError(f\"Not enough TN transcripts found. Found {len(tn_sample)}, required 250.\")\n",
    "\n",
    "# Save isoform IDs to files\n",
    "os.makedirs(filter_sets_dir, exist_ok=True)\n",
    "tp_file = os.path.join(filter_sets_dir, \"TP_list.txt\")\n",
    "tn_file = os.path.join(filter_sets_dir, \"TN_list.txt\")\n",
    "tp_sample['isoform'].to_csv(tp_file, index=False, header=False)\n",
    "tn_sample['isoform'].to_csv(tn_file, index=False, header=False)\n",
    "\n",
    "print(f\"Saved {len(tp_sample)} TP isoforms to {tp_file}\")\n",
    "print(f\"Saved {len(tn_sample)} TN isoforms to {tn_file}\")\n",
    "    \n",
    "# Exclusion file for variables that should't be useeed in ML filtering\n",
    "exclusion_file = os.path.join(filter_sets_dir, \"exclusion_list.txt\")\n",
    "with open(exclusion_file, 'w') as f:\n",
    "    f.write(\"all_canonical\\n\")\n",
    "    f.write(\"within_CAGE_peak\\n\")\n",
    "    f.write(\"within_polyA_site\\n\")\n",
    "    f.write(\"dist_to_CAGE_peak\\n\")\n",
    "    f.write(\"dist_to_polyA_site\\n\")\n",
    "print(f\"Exclusion list saved to {exclusion_file}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command:\n",
      "/usr/bin/time -v /home/fjetzinger/tools/SQANTI3-5.4/sqanti3_filter.py ml --sqanti_class /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_classification.txt --dir /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools --filter_isoforms /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_corrected.fasta --filter_gtf /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_corrected.gtf --threshold 0.7 --TP /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/filter_sets/TP_list.txt --TN /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/filter_sets/TN_list.txt --remove_columns /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/filter_sets/exclusion_list.txt --output h1_endo_chr8_isotools_filter --cpus 8\n",
      "Output:\n",
      "-------------------------------------------------\n",
      " \t SQANTI3 Machine Learning filter\n",
      "--------------------------------------------------\n",
      "CURRENT ML FILTER PARAMETERS:\n",
      " [1] \"sqanti_classif: /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_classification.txt\"\n",
      " [2] \"output: h1_endo_chr8_isotools_filter\"                                                                                                         \n",
      " [3] \"dir: /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools\"                                                   \n",
      " [4] \"percent_training: 0.8\"                                                                                                                        \n",
      " [5] \"threshold: 0.7\"                                                                                                                               \n",
      " [6] \"intrapriming: 60\"                                                                                                                             \n",
      " [7] \"force_fsm_in: FALSE\"                                                                                                                          \n",
      " [8] \"force_multi_exon: FALSE\"                                                                                                                      \n",
      " [9] \"intermediate_files: FALSE\"                                                                                                                    \n",
      "[10] \"max_class_size: 3000\"                                                                                                                         \n",
      "[11] \"help: FALSE\"                                                                                                                                  \n",
      "\n",
      "\tINITIAL ML CHECKS:\n",
      "Reading SQANTI3 *_classification.txt file...\n",
      "Checking data for mono and multi-exon transcripts...\n",
      " \t ***Note: ML filter can only be applied to multi-exon transcripts. \n",
      " \t 18947 multi-exon transcript isoforms found in SQ3 classification file.\n",
      "Checking input data for True Positive (TP) and True Negative (TN) sets...\n",
      "\tWarning message: \n",
      " \t Training set not provided -will be created from input data.\n",
      "Using Novel Not In Catalog non-canonical isoforms as True Negatives for training.\n",
      " \t - Total NNC non-canonical isoforms: 2745\n",
      "Using FSM Reference Match isoforms as True Positives for training\n",
      " \t - Total reference match isoforms (FSM subcategory): 534\n",
      "Excluding diff_to_* columns before model training to prevent overfitting!\n",
      " \t The following columns will be added to the column removal list: diff_to_gene_TSS \n",
      " \t The following columns will be added to the column removal list: diff_to_gene_TTS \n",
      " \t The following columns will be added to the column removal list: diff_to_TSS \n",
      " \t The following columns will be added to the column removal list: diff_to_TTS\n",
      "\n",
      "Balancing number of isoforms in TP and TN sets...\n",
      "\tMinimum set size: 534 transcripts.\n",
      "\tSampled 534 transcripts to define final TP and TN sets.\n",
      "Wrote generated TP and TN lists to files:\n",
      "\t/mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter_TP_list.txt\n",
      "\t/mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter_TN_list.txt\n",
      "-------------------------------------------------\n",
      "\tML DATA PREPARATION:\n",
      "Aggregating FL counts across samples (if more than one sample is provided)...\n",
      "Replacing NAs with appropriate values for ML...\n",
      "Handling factor columns...\n",
      "Handling integer columns...\n",
      "Removing variables with near-zero variance...\tRemoved columns: [1] \"chrom\"         \"n_indels\"      \"n_indels_junc\" \"ratio_exp\"    \n",
      "[5] \"coding\"        \"predicted_NMD\"\n",
      "\n",
      "Removing highly correlated features... (correlation threshold = 0.9).\n",
      "All correlations <= 0.9 \n",
      "\n",
      "\tList of removed features: \tNo features removed.\n",
      "-------------------------------------------------\n",
      "\tRANDOM FOREST ALGORITHM RUN:\n",
      "Creating positive and negative sets for classifier training and testing...\n",
      "Finished creating training data set.\n",
      "Partitioning data into training and test sets...\n",
      " \tProportion of the data to be used for training: 0.8\n",
      "Description of the training set:\n",
      " \tPositive and negative transcript isoforms in training set:\n",
      "   full-splice_match novel_not_in_catalog \n",
      "                 428                  428 \n",
      "\n",
      " \tPositive and negative transcript isoforms in test set:\n",
      "   full-splice_match novel_not_in_catalog \n",
      "                 106                  106 \n",
      "\n",
      "-------------------------------------------------\n",
      "Training Random Forest Classifier...\n",
      "\t***Note: this can take up to several hours.\n",
      "Pre-defined Random Forest parameters (supplied to caret::trainControl()):\t - Downsampling in training set (sampling = 'down').\t - 10x cross-validation (repeats = 10).\n",
      "\n",
      "Random forest training finished.\n",
      "Saved generated classifier to  /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter_randomforest.RData  file.\n",
      "-------------------------------------------------\n",
      "Random forest evaluation: applying classifier to test set...\n",
      "Test set evaluation results:------------------------------\n",
      "AUC, Sensitivity and Specificity on test set:      ROC      Sens      Spec \n",
      "0.9889640 0.8301887 1.0000000 \n",
      "\n",
      "Writing summary to testSet_summary.txt file.\n",
      "Confusion matrix:          Reference\n",
      "Prediction POS NEG\n",
      "       POS  88   0\n",
      "       NEG  18 106\n",
      "\n",
      "Writing confusion matrix and statistics to output files:\t testSet_confusionMatrix.txt\t testSet_stats.txt\n",
      "Global variable importance in Random Forest classifier:                         Overall\n",
      "min_cov               111.187183\n",
      "min_sample_cov        103.649470\n",
      "bite                   52.556916\n",
      "sd_cov                 28.458866\n",
      "length                 17.019546\n",
      "ratio_TSS              13.972372\n",
      "dist_to_CAGE_peak      13.558941\n",
      "exons                  12.274055\n",
      "dist_to_polyA_site     11.711867\n",
      "polyA_dist              9.212363\n",
      "perc_A_downstream_TTS   7.950733\n",
      "within_CAGE_peak        4.158515\n",
      "FSM_class               3.410537\n",
      "RTS_stage               2.766659\n",
      "within_polyA_site       2.218768\n",
      "polyA_motif_found       1.337199\n",
      "\n",
      "Variable importance table saved as classifier_variable-importance_table.txt\n",
      "Calculating and printing test set ROC curves...\n",
      "ROC curves saved to testSet_ROC_curve.pdf file. Includes:\t - ROC curve with unbalanced classes.\t - ROC curve with balanced classes.\n",
      "------------------------------------------------\n",
      "Applying Random Forest classifier to input dataset...\n",
      "Random forest prediction finished successfully!\n",
      "Random forest classification results:\n",
      "Negative Positive \n",
      "   11316     7631 \n",
      "\n",
      "-------------------------------------------------\n",
      "Applying intra-priming filter to our dataset.\n",
      "Intra-priming filtered transcripts:\n",
      "FALSE  TRUE \n",
      "22037   380 \n",
      "\n",
      " -------------------------------------------------\n",
      "Writing filter results to classification file...\n",
      "\tWrote filter results (ML and intra-priming) to new classification table:\n",
      "\th1_endo_chr8_isotools_filter_ML_result_classification.txt file.\n",
      "\tWrote isoform list (classified as non-artifacts by both ML and intra-priming\n",
      "\tfilters) to h1_endo_chr8_isotools_filter_inclusion-list.txt file\n",
      "-------------------------------------------------\n",
      "SUMMARY OF MACHINE LEARNING + INTRA-PRIMING FILTERS:\n",
      "Artifact  Isoform \n",
      "   11598    10819 \n",
      "\n",
      "-------------------------------------------------\n",
      "SQANTI3 ML filter finished successfully!\n",
      "-------------------------------------------------\n",
      "Errors:\n",
      "\n",
      "    ================================================     \n",
      "      ███████╗██╗██╗░░░░░████████╗███████╗██████╗░\n",
      "      ██╔════╝██║██║░░░░░╚══██╔══╝██╔════╝██╔══██╗\n",
      "      █████╗░░██║██║░░░░░░░░██║░░░█████╗░░██████╔╝\n",
      "      ██╔══╝░░██║██║░░░░░░░░██║░░░██╔══╝░░██╔══██╗\n",
      "      ██║░░░░░██║███████╗░░░██║░░░███████╗██║░░██║\n",
      "      ╚═╝░░░░░╚═╝╚══════╝░░░╚═╝░░░╚══════╝╚═╝░░╚═╝\n",
      "    ================================================\n",
      "    \n",
      "[WARNING:2025-05-14 15:20:28,478] Output directory /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools already exists. Overwriting!\n",
      "[INFO:2025-05-14 15:20:28,530] Write arguments to /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter_params.txt...\n",
      "[INFO:2025-05-14 15:20:28,551] Running SQANTI3 filtering...\n",
      "Loading required package: ggplot2\n",
      "Loading required package: lattice\n",
      "Setting levels: control = 1, case = 2\n",
      "Setting direction: controls > cases\n",
      "Setting levels: control = 1, case = 2\n",
      "Setting direction: controls < cases\n",
      "[INFO:2025-05-14 15:32:34,400] Output written to: /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter.filtered.fasta\n",
      "[INFO:2025-05-14 15:32:52,694] Output written to: /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter.filtered.gtf\n",
      "\tCommand being timed: \"/home/fjetzinger/tools/SQANTI3-5.4/sqanti3_filter.py ml --sqanti_class /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_classification.txt --dir /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools --filter_isoforms /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_corrected.fasta --filter_gtf /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/h1_endo_chr8_isotools/h1_endo_chr8_isotools_qc_corrected.gtf --threshold 0.7 --TP /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/filter_sets/TP_list.txt --TN /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/filter_sets/TN_list.txt --remove_columns /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/filter_sets/exclusion_list.txt --output h1_endo_chr8_isotools_filter --cpus 8\"\n",
      "\tUser time (seconds): 719.81\n",
      "\tSystem time (seconds): 21.78\n",
      "\tPercent of CPU this job got: 99%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 12:28.68\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 493564\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 208\n",
      "\tMinor (reclaiming a frame) page faults: 978678\n",
      "\tVoluntary context switches: 49822\n",
      "\tInvoluntary context switches: 6784\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 55920\n",
      "\tFile system outputs: 4536\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Run SQ3 Filter\n",
    "\n",
    "corrected_fasta = qc_dir + \"/\" + qc_prefix + \"_corrected.fasta\"\n",
    "corrected_gtf = qc_dir + \"/\" + qc_prefix + \"_corrected.gtf\"\n",
    "\n",
    "# Build the command\n",
    "\n",
    "# Default execution with automatic definition of TP and TN sets:\n",
    "# TP: all multi-exon Reference Match (RM, subcategory of FSM where TTS/TSS within 50bp); if <250, all FSM\n",
    "# TN: all multi-exon Novel Not in Catalog (NNC) that have at least one non-canonical junction; if <250, all NNC\n",
    "cmd_default = [\n",
    "    \"/usr/bin/time\", \"-v\",\n",
    "    f\"{sqanti_path}/sqanti3_filter.py\",         # SQANTI3 Filter script\n",
    "    \"ml\",                                       # Mode: Machine Learning\n",
    "    \"--sqanti_class\", classification_file,      # Classification file\n",
    "    \"--dir\", filter_dir,                        # Output Location\n",
    "    \"--filter_isoforms\", corrected_fasta,       # Corrected fasta (isoform sequences to filter)\n",
    "    \"--filter_gtf\", corrected_gtf,              # GTF to filter\n",
    "    \"--output\", filter_prefix,                  # Output prefix\n",
    "    \"--cpus\", str(n_cores),                     # Number of cores\n",
    "]\n",
    "\n",
    "# Command with manual definition of TP and TN sets\n",
    "cmd = [\n",
    "    \"/usr/bin/time\", \"-v\",\n",
    "    f\"{sqanti_path}/sqanti3_filter.py\",         # SQANTI3 Filter script\n",
    "    \"ml\",                                       # Mode: Machine Learning\n",
    "    \"--sqanti_class\", classification_file,      # Classification file\n",
    "    \"--dir\", filter_dir,                        # Output Location\n",
    "    \"--filter_isoforms\", corrected_fasta,       # Corrected fasta (isoform sequences to filter)\n",
    "    \"--filter_gtf\", corrected_gtf,              # GTF to filter\n",
    "    \"--threshold\", \"0.7\",                       # Threshold for Machine Learning\n",
    "    \"--TP\", tp_file,                            # TP list\n",
    "    \"--TN\", tn_file,                            # TN list\n",
    "    \"--remove_columns\", exclusion_file,         # Exclusion list\n",
    "    \"--output\", filter_prefix,                  # Output prefix\n",
    "    \"--cpus\", str(n_cores),                     # Number of cores\n",
    "]\n",
    "# Print the command for reference\n",
    "print(\"Running command:\")\n",
    "print(\" \".join(cmd))\n",
    "\n",
    "# Run the command\n",
    "result = subprocess.run(cmd, capture_output=True, text=True)\n",
    "\n",
    "# Print output and errors\n",
    "print(\"Output:\")\n",
    "print(result.stdout)\n",
    "print(\"Errors:\")\n",
    "print(result.stderr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Investigating SQANTI3 Filter Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Can you find answers to the following Questions in the SQANTI3 Filter report?\n",
    "\n",
    "Similarly to the report of the Quality Control process, SQANTI3 also provides a report for the Filter process. You can find it under `data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter_SQANTI3_filter_report.html`\n",
    "\n",
    "1. **Q1:**\n",
    "    \n",
    "    a. Q1a?\n",
    "\n",
    "    b. Q1b? \n",
    "\n",
    "2. **Q2:**\n",
    "\n",
    "    a. Q2a?\n",
    "\n",
    "    b. Q2b?\n",
    "\n",
    "3. **Q3:**\n",
    "\n",
    "    a. Q3a\n",
    "\n",
    "    b. Q3b\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div id=\"jb_3600609_buttons\"></div><div id=\"jb_3600609_igvcontainer\"></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "window.igv.MessageHandler.on({\"id\": \"jb_3600609\", \"command\": \"createBrowser\", \"data\": {\"reference\": {\"id\": \"hg38\", \"name\": \"Human (GRCH38/hg38)\", \"fastaPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa\", \"indexPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa.fai\"}, \"locus\": \"chr8:75,391,802-75,577,425\", \"tracks\": [{\"name\": \"IsoTools (after filter)\", \"path\": \"../../data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter.filtered.gtf\", \"format\": \"gtf\", \"type\": \"annotation\", \"displayMode\": \"SQUISHED\"}, {\"name\": \"Reference\", \"path\": \"../../data/sqanti_qc/gencode.v45.annotation_chr8/gencode.v45.annotation_chr8_qc_corrected.gtf\", \"format\": \"gtf\", \"type\": \"annotation\", \"displayMode\": \"SQUISHED\"}], \"id\": \"jb_3600609\"}})",
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "igv_browser_filter= igv_notebook.Browser(\n",
    "    {\n",
    "        \"reference\": {\n",
    "            \"id\": \"hg38\",\n",
    "            \"name\": \"Human (GRCH38/hg38)\",\n",
    "            \"fastaPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa\",\n",
    "            \"indexPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa.fai\"\n",
    "        },\n",
    "        \"locus\": \"chr8:75,391,802-75,577,425\",\n",
    "        \"tracks\": [\n",
    "            {\n",
    "                \"name\": \"IsoTools (after filter)\",\n",
    "                \"path\": \"../../data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter.filtered.gtf\",\n",
    "                \"format\": \"gtf\",\n",
    "                \"type\": \"annotation\",\n",
    "                \"displayMode\": \"SQUISHED\"\n",
    "            },\n",
    "            {\n",
    "            \"name\": \"Reference\",\n",
    "            \"path\": \"../../data/sqanti_qc/gencode.v45.annotation_chr8/gencode.v45.annotation_chr8_qc_corrected.gtf\",\n",
    "            \"format\": \"gtf\",\n",
    "            \"type\": \"annotation\",\n",
    "            \"displayMode\": \"SQUISHED\"\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SQANTI3 Rescue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO: Motivate SQANTI3 Rescue\n",
    "Copy some basic information, graphs, etc. about SQANTI from the docs and tutorials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    =====================================================\n",
      "      ██████╗░███████╗░██████╗░█████╗░██╗░░░██╗███████╗\n",
      "      ██╔══██╗██╔════╝██╔════╝██╔══██╗██║░░░██║██╔════╝\n",
      "      ██████╔╝█████╗░░╚█████╗░██║░░╚═╝██║░░░██║█████╗░░\n",
      "      ██╔══██╗██╔══╝░░░╚═══██╗██║░░██╗██║░░░██║██╔══╝░░\n",
      "      ██║░░██║███████╗██████╔╝╚█████╔╝╚██████╔╝███████╗\n",
      "      ╚═╝░░╚═╝╚══════╝╚═════╝░░╚════╝░░╚═════╝░╚══════╝\n",
      "    =====================================================\n",
      "    \n",
      "INFO:art_logger:\n",
      "    =====================================================\n",
      "      ██████╗░███████╗░██████╗░█████╗░██╗░░░██╗███████╗\n",
      "      ██╔══██╗██╔════╝██╔════╝██╔══██╗██║░░░██║██╔════╝\n",
      "      ██████╔╝█████╗░░╚█████╗░██║░░╚═╝██║░░░██║█████╗░░\n",
      "      ██╔══██╗██╔══╝░░░╚═══██╗██║░░██╗██║░░░██║██╔══╝░░\n",
      "      ██║░░██║███████╗██████╔╝╚█████╔╝╚██████╔╝███████╗\n",
      "      ╚═╝░░╚═╝╚══════╝╚═════╝░░╚════╝░░╚═════╝░╚══════╝\n",
      "    =====================================================\n",
      "    \n",
      "usage: sqanti3_rescue.py ml [-h] --filter_class FILTER_CLASS --refGTF REFGTF\n",
      "                            --refFasta REFFASTA\n",
      "                            [--rescue_isoforms RESCUE_ISOFORMS]\n",
      "                            [--rescue_gtf RESCUE_GTF] [-k REFCLASSIF]\n",
      "                            [--counts COUNTS] [-e {all,fsm,none}]\n",
      "                            [--mode {automatic,full}] [-q] [-o OUTPUT]\n",
      "                            [-d DIR] [-c CPUS] [-v]\n",
      "                            [-l {ERROR,WARNING,INFO,DEBUG}] [-r RANDOM_FOREST]\n",
      "                            [-j THRESHOLD]\n",
      "\n",
      "Rescue for ML-filtered transcriptomes.\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Required arguments:\n",
      "  --filter_class FILTER_CLASS\n",
      "                        SQANTI filter (ML or rules) output classification file.\n",
      "  --refGTF REFGTF       Full path to reference transcriptome GTF used when running SQANTI3 QC.\n",
      "  --refFasta REFFASTA   Full path to reference genome FASTA used when running SQANTI3 QC.\n",
      "\n",
      "Input options:\n",
      "  --rescue_isoforms RESCUE_ISOFORMS\n",
      "                        FASTA file output by SQANTI3 QC (*_corrected.fasta), i.e. the full long read transcriptome.\n",
      "  --rescue_gtf RESCUE_GTF\n",
      "                        GTF file output by SQANTI3 filter (*.filtered.gtf).\n",
      "  -k REFCLASSIF, --refClassif REFCLASSIF\n",
      "                        Full path to the classification file obtained when running SQANTI3 QC on the reference transcriptome.                        \n",
      "                        Mandatory when running the rescue on full mode\n",
      "  --counts COUNTS       Isoforms abundancy values: \"Isoform\" \t \"Count\". Column names may differ\n",
      "\n",
      "Customization options:\n",
      "  -e {all,fsm,none}, --rescue_mono_exonic {all,fsm,none}\n",
      "                        Whether or not to include mono-exonic artifacts in the rescue.                    \n",
      "                        Default: all\n",
      "  --mode {automatic,full}\n",
      "                        If 'automatic' (default), only automatic rescue of FSM artifacts will be performed.                     \n",
      "                        If 'full', rescue will include mapping of ISM, NNC and NIC artifacts to find potential replacement isoforms.\n",
      "  -q, --requant         Run requantification of the rescued isoforms.\n",
      "\n",
      "Output options:\n",
      "  -o OUTPUT, --output OUTPUT\n",
      "                        Prefix for output files.\n",
      "  -d DIR, --dir DIR     Directory for output files. Default: Directory where the script was run.\n",
      "\n",
      "Extra options:\n",
      "  -c CPUS, --cpus CPUS  Number of CPUs to use. Default: 4\n",
      "  -v, --version         Display program version number.\n",
      "  -l {ERROR,WARNING,INFO,DEBUG}, --log_level {ERROR,WARNING,INFO,DEBUG}\n",
      "                        Set the logging level INFO\n",
      "\n",
      "Machine Learning specific options:\n",
      "  -r RANDOM_FOREST, --random_forest RANDOM_FOREST\n",
      "                        Full path to the randomforest.RData object obtained when running the SQANTI3 ML filter.\n",
      "  -j THRESHOLD, --threshold THRESHOLD\n",
      "                        Machine learning probability threshold to filter elegible rescue targets (mapping hits).                     \n",
      "                        Default: 0.7\n"
     ]
    }
   ],
   "source": [
    "# How to run SQ3 Rescue?\n",
    "# see also: https://github.com/ConesaLab/SQANTI3/wiki/Running-SQANTI3-rescue\n",
    "!$sqanti_path/sqanti3_rescue.py ml --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running SQANTI3 QC to classify the reference transcriptome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command:\n",
      "/usr/bin/time -v /home/fjetzinger/tools/SQANTI3-5.4/sqanti3_qc.py --isoforms /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/gencode.v45.annotation_chr8.gtf --refGTF /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/gencode.v45.annotation_chr8.gtf --refFasta /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa --polyA_motif_list /home/fjetzinger/tools/SQANTI3-5.4/data/polyA_motifs/mouse_and_human.polyA_motif.txt --polyA_peak /mnt/c/data/lrgasp/homo_sapiens/reference/orthogonal/chr8/atlas.clusters.2.0.GRCh38.96_chr8.bed --coverage /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPSJ.out_chr8.tab --CAGE_peak /home/fjetzinger/tools/SQANTI3-5.4/data/ref_TSS_annotation/human.refTSS_v3.1.hg38.bed --SR_bam /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPAligned.sortedByCoord.out_chr8.bam --output gencode.v45.annotation_chr8_qc --dir /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/gencode.v45.annotation_chr8 --skipORF --cpus 8\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output:\n",
      "\n",
      "Errors:\n",
      "\n",
      "    =====================\n",
      "      ░██████╗░░█████╗░\n",
      "      ██╔═══██╗██╔══██╗\n",
      "      ██║██╗██║██║░░╚═╝\n",
      "      ╚██████╔╝██║░░██╗\n",
      "      ░╚═██╔═╝░╚█████╔╝\n",
      "      ░░░╚═╝░░░░╚════╝░\n",
      "    =====================\n",
      "    \n",
      "[INFO:2025-05-14 15:32:58,154] Write arguments to /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/gencode.v45.annotation_chr8/gencode.v45.annotation_chr8_qc.qc_params.txt...\n",
      "[INFO:2025-05-14 15:32:58,225] Initialising QC pipeline.\n",
      "[INFO:2025-05-14 15:32:58,226] Parsing provided files\n",
      "[INFO:2025-05-14 15:32:58,259] Reading genome fasta /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa\n",
      "[INFO:2025-05-14 15:33:08,794] **** Correcting sequences\n",
      "[INFO:2025-05-14 15:33:08,795] Correcting fasta\n",
      "[INFO:2025-05-14 15:33:08,796] Skipping aligning of sequences because GTF file was provided.\n",
      "[INFO:2025-05-14 15:33:15,329] Indels will be not calculated since you ran SQANTI3 without alignment step (SQANTI3 with gtf format as transcriptome input).\n",
      "[INFO:2025-05-14 15:33:31,637] **** Predicting ORF sequences...\n",
      "[WARNING:2025-05-14 15:33:31,643] Skipping ORF prediction because user requested it. All isoforms will be non-coding!\n",
      "[WARNING:2025-05-14 15:33:31,643] All input isoforms were predicted as non-coding\n",
      "[INFO:2025-05-14 15:33:31,644] **** Parsing Reference Transcriptome....\n",
      "[INFO:2025-05-14 15:33:35,047] **** Parsing Isoforms.\n",
      "[INFO:2025-05-14 15:33:35,477] Input pattern: /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPSJ.out_chr8.tab.\n",
      "The following files found and to be read as junctions:\n",
      "/mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPSJ.out_chr8.tab\n",
      "[INFO:2025-05-14 15:33:35,814] 10577 junctions read. 14 junctions added to both strands because no strand information from STAR.\n",
      "[INFO:2025-05-14 15:33:35,815] Using provided BAM files for calculating TSS ratio\n",
      "[INFO:2025-05-14 15:33:38,496] BAM files identified: ['/mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPAligned.sortedByCoord.out_chr8.bam']\n",
      "[INFO:2025-05-14 15:34:21,740] **** Reading CAGE Peak data.\n",
      "[INFO:2025-05-14 15:34:22,682] **** Reading polyA Peak data.\n",
      "[INFO:2025-05-14 15:34:23,060] **** Reading PolyA motif list.\n",
      "[INFO:2025-05-14 15:34:23,061] **** Performing Classification of Isoforms\n",
      "[INFO:2025-05-14 15:35:28,965] Number of classified isoforms: 10272\n",
      "[INFO:2025-05-14 15:35:28,999] After classify fsm: 10272\n",
      "[INFO:2025-05-14 15:35:29,000] Full-length read abundance files not provided.\n",
      "[INFO:2025-05-14 15:35:29,000] RT-switching computation\n",
      "[INFO:2025-05-14 15:35:30,511] After RTS classificaion: 10272\n",
      "[INFO:2025-05-14 15:35:30,513] **** Adding TSS ratio data.\n",
      "[INFO:2025-05-14 15:35:30,543] Isoforms expression files not provided.\n",
      "[INFO:2025-05-14 15:35:31,419] Writing output files\n",
      "[INFO:2025-05-14 15:35:41,273] **** Generating SQANTI3 report.\n",
      "[INFO:2025-05-14 15:36:19,483] Removing temporary files.\n",
      "[INFO:2025-05-14 15:36:19,492] SQANTI3 complete in 201.26847322599497 sec.\n",
      "\tCommand being timed: \"/home/fjetzinger/tools/SQANTI3-5.4/sqanti3_qc.py --isoforms /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/gencode.v45.annotation_chr8.gtf --refGTF /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/gencode.v45.annotation_chr8.gtf --refFasta /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa --polyA_motif_list /home/fjetzinger/tools/SQANTI3-5.4/data/polyA_motifs/mouse_and_human.polyA_motif.txt --polyA_peak /mnt/c/data/lrgasp/homo_sapiens/reference/orthogonal/chr8/atlas.clusters.2.0.GRCh38.96_chr8.bed --coverage /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPSJ.out_chr8.tab --CAGE_peak /home/fjetzinger/tools/SQANTI3-5.4/data/ref_TSS_annotation/human.refTSS_v3.1.hg38.bed --SR_bam /mnt/c/data/lrgasp/homo_sapiens/endodermal/illumina_cdna/chr8/ENCFF498FDF_ENCFF181VTPAligned.sortedByCoord.out_chr8.bam --output gencode.v45.annotation_chr8_qc --dir /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/gencode.v45.annotation_chr8 --skipORF --cpus 8\"\n",
      "\tUser time (seconds): 135.66\n",
      "\tSystem time (seconds): 39.62\n",
      "\tPercent of CPU this job got: 84%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 3:27.14\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 601812\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 1\n",
      "\tMinor (reclaiming a frame) page faults: 321621\n",
      "\tVoluntary context switches: 131748\n",
      "\tInvoluntary context switches: 28672\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 336\n",
      "\tFile system outputs: 14696\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# %%script echo Skipping SQANTI3 QC on reference\n",
    "# Run SQ3 QC on reference with the same orthogonal data\n",
    "cmd = [\n",
    "    \"/usr/bin/time\", \"-v\",\n",
    "    f\"{sqanti_path}/sqanti3_qc.py\",         # SQANTI3 QC script\n",
    "    \"--isoforms\", ref_gtf,                  # GTF to Quality Control\n",
    "    \"--refGTF\", ref_gtf,                    # Reference GTF\n",
    "    \"--refFasta\", ref_genome,               # Reference Genome\n",
    "    \"--polyA_motif_list\", polyA_motifs,     # PolyA Motif List\n",
    "    \"--polyA_peak\", polyA_peaks,            # PolyA Peaks\n",
    "    # \"--fl_count\", counts,                 # Counts file\n",
    "    \"--coverage\", splice_junctions,         # SJ file\n",
    "    \"--CAGE_peak\", cage,                    # CAGE Peak file\n",
    "    \"--SR_bam\", sr_bam,                     # SR BAM file\n",
    "    \"--output\", ref_qc_prefix,              # Output Prefix\n",
    "    \"--dir\", ref_qc_dir,                    # Output Location\n",
    "    \"--skipORF\",                            # Skip ORF Prediction (takes longer)\n",
    "    \"--cpus\", str(n_cores)                  # Number of Threads\n",
    "]\n",
    "\n",
    "# Print the command for reference\n",
    "print(\"Running command:\")\n",
    "print(\" \".join(cmd))\n",
    "\n",
    "# Run the command\n",
    "result = subprocess.run(cmd, capture_output=True, text=True)\n",
    "\n",
    "# Print output and errors\n",
    "print(\"Output:\")\n",
    "print(result.stdout)\n",
    "print(\"Errors:\")\n",
    "print(result.stderr)\n",
    "\n",
    "ref_classification = os.path.join(ref_qc_dir, f\"{ref_qc_prefix}_classification.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running SQANTI3 Rescue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command:\n",
      "/usr/bin/time -v /home/fjetzinger/tools/SQANTI3-5.4/sqanti3_rescue.py ml --rescue_isoforms /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter.filtered.fasta --rescue_gtf /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter.filtered.gtf --filter_class /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter_ML_result_classification.txt --refGTF /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/gencode.v45.annotation_chr8.gtf --refFasta /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa --refClassif /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/gencode.v45.annotation_chr8/gencode.v45.annotation_chr8_qc_classification.txt --random_forest /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter_randomforest.RData --threshold 0.7 --mode full --output h1_endo_chr8_isotools_rescue --dir /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_rescue/h1_endo_chr8_isotools --cpus 8\n",
      "Output:\n",
      "\n",
      "Errors:\n",
      "\n",
      "    =====================================================\n",
      "      ██████╗░███████╗░██████╗░█████╗░██╗░░░██╗███████╗\n",
      "      ██╔══██╗██╔════╝██╔════╝██╔══██╗██║░░░██║██╔════╝\n",
      "      ██████╔╝█████╗░░╚█████╗░██║░░╚═╝██║░░░██║█████╗░░\n",
      "      ██╔══██╗██╔══╝░░░╚═══██╗██║░░██╗██║░░░██║██╔══╝░░\n",
      "      ██║░░██║███████╗██████╔╝╚█████╔╝╚██████╔╝███████╗\n",
      "      ╚═╝░░╚═╝╚══════╝╚═════╝░░╚════╝░░╚═════╝░╚══════╝\n",
      "    =====================================================\n",
      "    \n",
      "INFO:art_logger:\n",
      "    =====================================================\n",
      "      ██████╗░███████╗░██████╗░█████╗░██╗░░░██╗███████╗\n",
      "      ██╔══██╗██╔════╝██╔════╝██╔══██╗██║░░░██║██╔════╝\n",
      "      ██████╔╝█████╗░░╚█████╗░██║░░╚═╝██║░░░██║█████╗░░\n",
      "      ██╔══██╗██╔══╝░░░╚═══██╗██║░░██╗██║░░░██║██╔══╝░░\n",
      "      ██║░░██║███████╗██████╔╝╚█████╔╝╚██████╔╝███████╗\n",
      "      ╚═╝░░╚═╝╚══════╝╚═════╝░░╚════╝░░╚═════╝░╚══════╝\n",
      "    =====================================================\n",
      "    \n",
      "[INFO:2025-05-14 16:17:25,429] Running SQANTI3 rescue pipeline version 5.4\n",
      "[INFO:2025-05-14 16:17:25,438] -----------------------------------------------------\n",
      "[INFO:2025-05-14 16:17:25,439]   Initializing SQANTI3 rescue pipeline in full mode  \n",
      "[INFO:2025-05-14 16:17:25,440] -----------------------------------------------------\n",
      "[INFO:2025-05-14 16:17:25,441] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:17:25,442]         Reading filter classification file        \n",
      "[INFO:2025-05-14 16:17:25,443] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:17:25,685] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:17:25,686]            Performing automatic rescue            \n",
      "[INFO:2025-05-14 16:17:25,687] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:17:25,894] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:17:25,895]             Automatic rescue completed            \n",
      "[INFO:2025-05-14 16:17:25,895] --------------------------------------------------\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'ont', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'hgnc_id', 'protein_id', 'ccdsid']\n",
      "[INFO:2025-05-14 16:17:31,762] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:17:31,763]     Preparation of files for artifact mapping:    \n",
      "[INFO:2025-05-14 16:17:31,764] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:17:31,765] Creating reference transcriptome FASTA from provided GTF (--refGTF).\n",
      "[INFO:2025-05-14 16:17:45,116] Reference transcriptome FASTA was saved to /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_rescue/h1_endo_chr8_isotools/gencode.v45.annotation_chr8.fasta\n",
      "[INFO:2025-05-14 16:17:45,117] gffread command used:\n",
      "[INFO:2025-05-14 16:17:45,118] gffread -w /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_rescue/h1_endo_chr8_isotools/gencode.v45.annotation_chr8.fasta -g /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/gencode.v45.annotation_chr8.gtf\n",
      "[INFO:2025-05-14 16:17:45,119] Filtering reference transcriptome FASTA to only rescue targets.\n",
      "[INFO:2025-05-14 16:17:48,508] Filtering supplied long read transcriptome FASTA (--isoforms) to only include rescue targets...\n",
      "[INFO:2025-05-14 16:17:56,872] Creating rescue candidate FASTA from supplied long read transcriptome fasta (--isoforms)...\n",
      "[INFO:2025-05-14 16:18:02,950] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:18:02,951]      Artifact mapping (candidates vs targets)     \n",
      "[INFO:2025-05-14 16:18:02,952] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:18:02,954] Mapping rescue candidates to rescue targets with minimap2...\n",
      "[INFO:2025-05-14 16:18:07,689] Minimap2 results were saved to /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_rescue/h1_endo_chr8_isotools/h1_endo_chr8_isotools_rescue_mapped_rescue.sam\n",
      "[INFO:2025-05-14 16:18:07,691] Building candidate-target table of mapping hits...\n",
      "[INFO:2025-05-14 16:18:08,116] Mapping hit table was saved to /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_rescue/h1_endo_chr8_isotools/h1_endo_chr8_isotools_rescue_rescue_mapping_hits.tsv\n",
      "[INFO:2025-05-14 16:18:08,235] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:18:08,237]          Rescue-by-mapping for ML filter          \n",
      "[INFO:2025-05-14 16:18:08,238] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:18:08,239] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:18:08,241]                ML rescue selected!                \n",
      "[INFO:2025-05-14 16:18:08,242] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:18:08,244] Running pre-trained random forest on reference transcriptome classification file.\n",
      "[INFO:2025-05-14 16:18:17,996] Running rescue-by-mapping for ML filter.\n",
      "[INFO:2025-05-14 16:18:20,433] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:18:20,434]          Rescue ml finished successfully!         \n",
      "[INFO:2025-05-14 16:18:20,435] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:18:20,435] Final rescued transcript list written to file: /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_rescue/h1_endo_chr8_isotools/h1_endo_chr8_isotools_rescue_rescue_inclusion-list.tsv\n",
      "[INFO:2025-05-14 16:18:20,436] Adding rescued transcripts to provided SQ3 filtered GTF.\n",
      "[INFO:2025-05-14 16:18:33,092] Added rescued reference transcripts to provided GTF (/mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter.filtered.gtf )\n",
      "[INFO:2025-05-14 16:18:33,100] Final output GTF written to file:  /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_rescue/h1_endo_chr8_isotools/h1_endo_chr8_isotools_rescue_rescued.gtf \n",
      "[INFO:2025-05-14 16:18:33,125] --------------------------------------------------\n",
      "[INFO:2025-05-14 16:18:33,127]           Rescue finished successfully!           \n",
      "[INFO:2025-05-14 16:18:33,131] --------------------------------------------------\n",
      "\tCommand being timed: \"/home/fjetzinger/tools/SQANTI3-5.4/sqanti3_rescue.py ml --rescue_isoforms /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter.filtered.fasta --rescue_gtf /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter.filtered.gtf --filter_class /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter_ML_result_classification.txt --refGTF /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/gencode.v45.annotation_chr8.gtf --refFasta /mnt/c/data/lrgasp/homo_sapiens/reference/chr8/GRCh38.primary_assembly.genome_chr8.fa --refClassif /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_qc/gencode.v45.annotation_chr8/gencode.v45.annotation_chr8_qc_classification.txt --random_forest /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_filter/h1_endo_chr8_isotools/h1_endo_chr8_isotools_filter_randomforest.RData --threshold 0.7 --mode full --output h1_endo_chr8_isotools_rescue --dir /mnt/c/Users/jetzi/other_repos/summer_school/data/sqanti_rescue/h1_endo_chr8_isotools --cpus 8\"\n",
      "\tUser time (seconds): 27.22\n",
      "\tSystem time (seconds): 9.48\n",
      "\tPercent of CPU this job got: 49%\n",
      "\tElapsed (wall clock) time (h:mm:ss or m:ss): 1:14.86\n",
      "\tAverage shared text size (kbytes): 0\n",
      "\tAverage unshared data size (kbytes): 0\n",
      "\tAverage stack size (kbytes): 0\n",
      "\tAverage total size (kbytes): 0\n",
      "\tMaximum resident set size (kbytes): 819412\n",
      "\tAverage resident set size (kbytes): 0\n",
      "\tMajor (requiring I/O) page faults: 1770\n",
      "\tMinor (reclaiming a frame) page faults: 288914\n",
      "\tVoluntary context switches: 93719\n",
      "\tInvoluntary context switches: 142\n",
      "\tSwaps: 0\n",
      "\tFile system inputs: 396688\n",
      "\tFile system outputs: 0\n",
      "\tSocket messages sent: 0\n",
      "\tSocket messages received: 0\n",
      "\tSignals delivered: 0\n",
      "\tPage size (bytes): 4096\n",
      "\tExit status: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "filter_classification = os.path.join(filter_dir, f\"{filter_prefix}_ML_result_classification.txt\")\n",
    "filtered_fasta = os.path.join(filter_dir, f\"{filter_prefix}.filtered.fasta\")\n",
    "filtered_gtf = os.path.join(filter_dir, f\"{filter_prefix}.filtered.gtf\")\n",
    "random_forest_model = os.path.join(filter_dir, f\"{filter_prefix}_randomforest.RData\")\n",
    "\n",
    "# Run SQ3 Rescue\n",
    "cmd = [\n",
    "    \"/usr/bin/time\", \"-v\",\n",
    "    f\"{sqanti_path}/sqanti3_rescue.py\",         # SQANTI3 Rescue script\n",
    "    \"ml\",                                       # Filter Mode: Machine Learning\n",
    "    \"--rescue_isoforms\", filtered_fasta,        # Filtered fasta\n",
    "    \"--rescue_gtf\", filtered_gtf,               # Filtered GTF\n",
    "    \"--filter_class\", filter_classification,    # Filter Classification\n",
    "    \"--refGTF\", ref_gtf,                        # Reference GTF\n",
    "    \"--refFasta\", ref_genome,                   # Reference Genome\n",
    "    \"--refClassif\", ref_classification,         # Reference Classification\n",
    "    # \"--requant\",                              # Requantify\n",
    "    # \"--counts\", counts,                       # Counts file\n",
    "    \"--random_forest\", random_forest_model,     # Random Forest Model from Filter\n",
    "    \"--threshold\", \"0.7\",                       # Threshold for Machine Learning\n",
    "    \"--mode\", \"full\",                           # Rescue Mode: Full (extend rescue to non-FSM isoforms)\n",
    "    \"--output\", rescue_prefix,                  # Output Prefix\n",
    "    \"--dir\", rescue_dir,                        # Output Location\n",
    "    \"--cpus\", str(n_cores)                      # Number of Threads\n",
    "]\n",
    "\n",
    "# Print the command for reference\n",
    "print(\"Running command:\")\n",
    "print(\" \".join(cmd))\n",
    "\n",
    "# Run the command\n",
    "result = subprocess.run(cmd, capture_output=True, text=True)\n",
    "\n",
    "# Print output and errors\n",
    "print(\"Output:\")\n",
    "print(result.stdout)\n",
    "print(\"Errors:\")\n",
    "print(result.stderr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Investigating SQANTI3 Rescue Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Can you find answers to the following Questions in the SQANTI3 Rescue results?\n",
    "\n",
    "While SQANTI3 Rescue does not provide a report, we can explore the results to find answers to the following questions.\n",
    "\n",
    "1. **Q1:**\n",
    "    \n",
    "    a. Q1a?\n",
    "\n",
    "    b. Q1b? \n",
    "\n",
    "2. **Q2:**\n",
    "\n",
    "    a. Q2a?\n",
    "\n",
    "    b. Q2b?\n",
    "\n",
    "3. **Q3:**\n",
    "\n",
    "    a. Q3a\n",
    "\n",
    "    b. Q3b\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div id=\"jb_3750949_buttons\"></div><div id=\"jb_3750949_igvcontainer\"></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "window.igv.MessageHandler.on({\"id\": \"jb_3750949\", \"command\": \"createBrowser\", \"data\": {\"reference\": {\"id\": \"hg38\", \"name\": \"Human (GRCH38/hg38)\", \"fastaPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa\", \"indexPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa.fai\"}, \"locus\": \"chr8:28701580-28753690\", \"tracks\": [{\"name\": \"IsoTools (after rescue)\", \"path\": \"../../data/sqanti_rescue/h1_endo_chr8_isotools/h1_endo_chr8_isotools_rescue_rescued.gtf\", \"format\": \"gtf\", \"type\": \"annotation\", \"displayMode\": \"SQUISHED\"}, {\"name\": \"Reference\", \"path\": \"../../data/sqanti_qc/gencode.v45.annotation_chr8/gencode.v45.annotation_chr8_qc_corrected.gtf\", \"format\": \"gtf\", \"type\": \"annotation\", \"displayMode\": \"SQUISHED\"}], \"id\": \"jb_3750949\"}})",
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "igv_browser_rescue= igv_notebook.Browser(\n",
    "    {\n",
    "        \"reference\": {\n",
    "            \"id\": \"hg38\",\n",
    "            \"name\": \"Human (GRCH38/hg38)\",\n",
    "            \"fastaPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa\",\n",
    "            \"indexPath\": \"../../data/reference/GRCh38.primary_assembly.genome_chr8.fa.fai\"\n",
    "        },\n",
    "        \"locus\": \"chr8:28701580-28753690\",\n",
    "        \"tracks\": [\n",
    "            {\n",
    "                \"name\": \"IsoTools (after rescue)\",\n",
    "                \"path\": \"../../data/sqanti_rescue/h1_endo_chr8_isotools/h1_endo_chr8_isotools_rescue_rescued.gtf\",\n",
    "                \"format\": \"gtf\",\n",
    "                \"type\": \"annotation\",\n",
    "                \"displayMode\": \"SQUISHED\"\n",
    "            },\n",
    "            {\n",
    "            \"name\": \"Reference\",\n",
    "            \"path\": \"../../data/sqanti_qc/gencode.v45.annotation_chr8/gencode.v45.annotation_chr8_qc_corrected.gtf\",\n",
    "            \"format\": \"gtf\",\n",
    "            \"type\": \"annotation\",\n",
    "            \"displayMode\": \"SQUISHED\"\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Where do we go from here?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After using SQANTI3 to perform Quality Control on a custom transcriptome, Filter out spurious transcripts, and Rescue related reference transcripts, further downstream analyses can be conducted.\n",
    "\n",
    "Some examples of further analyses:\n",
    "\n",
    "* Differential Gene Expression; Differential Transcript Expression/Usage; Differential Exon Expression/Usage\n",
    "\n",
    "* Functional Annotation, Enrichment Analysis, Pathway Analysis"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
