# Day 4 Agenda

## Goals
* 
*


## Timetable
| Time          | Activity                | Details                                                                                                                                       |
| --------------------- | --------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| 09:00 - 09:10 | Introduction & Objectives         | <u>Eamon</u>: Introduction                                                                                                     |
| 09:10 - 09:25 | Long Read Single-cell Library Preparation    | <u>Fran</u>: Library preparation                                                                                                                |
| 09:25 - 10:00 | Preprocessing and QC    | <u>Fran</u>: Introduce commercial pipelines, Steps (From UBAM to quantification)(Point to the tool for each step), recommendations                                                                                          |
| 10:00 - 10:30 | Q&A Break       | Break                                                                                                                                                                    |
| 10:30 - 11:00 | Tertiary analysis   | <u>Eamon</u>: Filtering, QC, cell typing, gene level analysis, isoform level analysis.                                                  |
| 11:00 - 12:30 | Hands-on    | <u>Eamon and Fran</u>: Quantification exercise: Commands to map, checking output files (highlight differences between bulk output). Tertiary analysis exercise: QC, Standard Workflow for single cell, Differential Isoform Usage/Isoform switches.   |
| 12:30 - 13:30 | Lunch Break | Lunch and informal networking                                                                                                       |
| 13:30 - 13:40 | Introduction & Objectives | <u>Carmen</u>: Welcome participants, overview of the workshop goals and structure. Learning outcomes and expectations.                                                                                                                     |
| 13:40 - 14:00 | Introduction to Metatranscriptomics  |<u>Carmen</u>: What is metatranscriptomics? Difference with classical transcriptomics, applications, short vs long reads, challenges (rRNA contamination, read orientation issues, incomplete annotations)                                                                           |
| 14:00 - 15:00 | Metatranscriptomics workflow  | <u>Carmen</u>: Understand all steps of the metatranscriptomic workflow: Experimental design, RNA extraction and purification, library preparation  (prokaryotic vs eukaryotic), RNASeq, Filtering and QC, Downstream analysis (mapping and quantification, taxonomic classification, functional annotation).        |
| 15:00 - 15:30 | Coffee Break & Group Discussion    | Break and open discussion |
| 15:30 - 15:45 | Practical Setup & Dataset Overview    | <u>Carmen</u>: Description of environmental samples, sequencing platform used, introduction to data files, preview of analysis tasks.                                                                                                                                                       |
| 15:45 - 16:15 | Task 1: Mapping | <u>Carmen</u>: Mapping metatranscriptomic reads to MAGs, reference genomes and metagenomic readset. Handling alignment files (different types of alignments, CIGAR and MD, orientation challenges).                                                               |
| 16:15 - 16:30 | Q&A Break | Open discussion, troubleshooting, clarification of Task1.                                                                                                                                                      |
| 16:30 - 17:00 | Task 2: Taxonomic Diversity adn Functional Activity | <u>Carmen</u>: Taxonomic analysis: Species richness and diversity. Functional analysis: Determine which biological processes or pathways are actively expressed in different environments. 


## Learning Outcomes
* 
* 


## Materials
* (here we will put links to the slides)


## Data
* 


## Recommended Bibliography
*
*
*
*
