---
layout: page
title: Day 4
schemadotorg:
  "@context": http://schema.org/
  "@type": CreativeWork
  "genre": TrainingMaterial
  isPartOf:
      url: "https://longtrec.github.io/summer_school/"
      name: "LongTREC summer school - long-read transcriptomics"
---

<img src="../assets/logos/LongTREC_logo_FINAL.png" width="200" />

# Day 4 Agenda

## Goals
**Morning session: Single-cell transcriptomics with long-reads** 
* Learn the differences between bulk experiments and single-cell experiments (library preparation and data)
* Understand the different steps involved in going from Long Sequencing Reads to gene and isoform matrices.
* Learn the basics of tertiary analysis for single cell data with focus on long-read specific steps.


**Afternoon session: Metatranscriptomics** 
* Provide participants with a comprehensive understanding of the **metatranscriptomics workflow**
* Introduce practical skills for **analyzing long-read metatranscriptomic data**
* Highlight the **challenges and considerations** unique to environmental metatranscriptomics


## Timetable

| Time          | Activity                | Details                                                                                                                                       |
| --------------------- | --------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| 09:00 - 09:15 | Introduction & Objectives         | <u>Eamon</u>: Introduction                                                                                                     |
| 09:15 - 09:30 | Long-Read Single-cell Library Preparation    | <u>Fran</u>:  Differences between bulk library preparation and single-cell library preparation.                                                                                                            |                                                                                                                                                                  |
| 09:30 - 10:00 | Preprocessing and QC    | <u>Fran</u>: Commercial analysis pipelines available, overview of the analysis steps and additional tools available.                                                                                          |
| 10:00 - 10:30 | Coffee Break       | Break                                                                                                                                                                    |
| 10:30 - 11:00 | Tertiary analysis   | <u>Eamon</u>: Filtering, QC, cell typing, gene level analysis, isoform level analysis.                                                  |
| 11:00 - 12:30 | Hands-on    | <u>Eamon and Fran</u>: 1) Iso-Seq pipeline overview: commands, intermediate files and report. 2) Tertiary analysis: exploring differential isoform usage across different cell types.  |
| 12:30 - 13:30 | Lunch Break | Lunch and informal networking                                                                                                       |
| 13:30 - 13:40 | Introduction & Objectives | <u>Carmen</u>: Welcome participants, overview of the workshop goals and structure. Learning outcomes and expectations.                                                                                                                     |
| 13:40 - 14:00 | Introduction to Metatranscriptomics  |<u>Carmen</u>: What is metatranscriptomics? Difference with classical transcriptomics, applications, short vs long reads, challenges (rRNA contamination, read orientation issues, incomplete annotations)                                                                           |
| 14:00 - 15:00 | Metatranscriptomics workflow  | <u>Carmen</u>: Understand all steps of the metatranscriptomic workflow: Experimental design, RNA extraction and purification, library preparation  (prokaryotic vs eukaryotic), RNASeq, Filtering and QC, Downstream analysis (mapping and quantification, taxonomic classification, functional annotation).        |
| 15:00 - 15:30 | Coffee Break & Group Discussion    | Break and open discussion |
| 15:30 - 15:45 | Practical Setup & Dataset Overview    | <u>Carmen</u>: Description of environmental samples, sequencing platform used, introduction to data files, preview of analysis tasks.                                                                                                                                                       |
| 15:45 - 16:15 | Task 1:  | <u>Carmen</u>: Pre-processing metatranscriptomic reads.                                                               |
| 16:15 - 16:30 | Q&A Break | Open discussion, troubleshooting, clarification of Task1.                                                                                                                                                      |
| 16:30 - 17:00 | Task 2 & 3: Mapping, Taxonomic assignation and Functional analysis | <u>Carmen</u>: Mapping metatranscriptomic reads, handling alignment files, perform taxonomic assignation and functional analysis.


## Learning Outcomes
**Morning session: Single-cell and spatial transcriptomics with long reads** 
* Identify and understand differences between bulk and single-cell protocols
* Understand single-cell analysis pipeline steps


**Afternoon session: Metatranscriptomics** 
* Explore applications of metatranscriptomics in environmental studies
* Understand what is metatranscriptomics workflow from RNA extraction to data interpretation
* Recognize challenges and biases regarding metatranscriptomics
* Perform quality control of raw metatranscriptomic reads
* Explore taxonomic diversity in different communities and characterise active functions


## Materials
* (here we will put links to the slides)


## Data
* 


## Recommended Bibliography
* Gupta, P., O’Neill, H., Wolvetang, E.J., Chatterjee, A., Gupta, I., 2024. Advances in single-cell long-read sequencing technologies. NAR Genomics and Bioinformatics 6, lqae047. https://doi.org/10.1093/nargab/lqae047
* Heumos, L., Schaar, A.C., Lance, C., Litinetskaya, A., Drost, F., Zappia, L., Lücken, M.D., Strobl, D.C., Henao, J., Curion, F., Schiller, H.B., Theis, F.J., 2023. Best practices for single-cell analysis across modalities. Nat Rev Genet 24, 550–572. https://doi.org/10.1038/s41576-023-00586-w
*
*

### Back

Back to [main page](../index.md).
